File: EscapeNine-endless-/EscapeNine-endless-/Models/Skill.swift
//
//  Skill.swift
//  EscapeNine-endless-
//
//  Created by 吉留徹 on 2025/11/14.
//

import Foundation

enum SkillType: String {
    case dash = "dash"
    case diagonal = "diagonal"
    case invisible = "invisible"
    case bind = "bind"
    
    var name: String {
        switch self {
        case .dash: return "ダッシュ"
        case .diagonal: return "斜め移動"
        case .invisible: return "透明化"
        case .bind: return "拘束"
        }
    }
    
    var description: String {
        switch self {
        case .dash: return "2マス移動できる"
        case .diagonal: return "斜め方向に移動可能"
        case .invisible: return "鬼に当たっても無敵"
        case .bind: return "鬼を1ターン停止させる"
        }
    }
}

struct Skill {
    let type: SkillType
    let name: String
    let description: String
    let maxUsage: Int
}

File: unity/EscapeNine/Assets/Scripts/Core/Skill.cs
// Skill.cs
// Swift 正本からの忠実移植: Models/Skill.swift

namespace EscapeNine.Core
{
    public static class SkillTypeInfo
    {
        public static string Name(this SkillType type) => type switch
        {
            SkillType.Dash => "ダッシュ",
            SkillType.Diagonal => "斜め移動",
            SkillType.Invisible => "透明化",
            SkillType.Bind => "拘束",
            SkillType.Shield => "盾ガード",
            _ => ""
        };

        public static string Description(this SkillType type) => type switch
        {
            SkillType.Dash => "2マス移動できる",
            SkillType.Diagonal => "斜め方向に移動可能",
            SkillType.Invisible => "鬼に当たっても無敵",
            SkillType.Bind => $"鬼を{GameConfig.BindDurationTurns}ターン停止させる",
            SkillType.Shield => "次の衝突を1回無効化する",
            _ => ""
        };
    }

    /// <summary>スキル定義。Swift: struct Skill</summary>
    public readonly struct Skill
    {
        public readonly SkillType Type;
        public readonly string Name;
        public readonly string Description;
        public readonly int MaxUsage;

        public Skill(SkillType type, string name, string description, int maxUsage)
        {
            Type = type;
            Name = name;
            Description = description;
            MaxUsage = maxUsage;
        }
    }
}
